### ISSUE TITLE
Interpreter shutdown support

### ISSUE THREAD

**Issue by thorink** [pybind/pybind11#749]
In our project we have to perform some shutdown tasks. At the moment we use the Python API function `Py_AtExit` directly (Actually a pybind11 wrapper for this would be nice. Should I work on a MR for this?). I can attach a function that gets called when the python interpreter exits.

Now, when I tidy up remaining `py::object`s in the callback attached to `Py_AtExit`, the `handle::dec_ref` calls the Python API in the destructor of `py::object`. This is not allowed as stated in the CPython docs (https://docs.python.org/3.5/c-api/sys.html#c.Py_AtExit):

> Since Python’s internal finalization will have completed before the cleanup function, no Python APIs should be called by func.

Does anyone has a nice solution for this?

**Comment by thorink** [comment]
Reading the release notes of v2.1 I found out about this module destructor trick ([link](http://pybind11.readthedocs.io/en/latest/advanced/misc.html?highlight=capsule#module-destructors)):

```c++
auto cleanup_callback = []() {
    // perform cleanup here -- this function is called with the GIL held
};

m.add_object("_cleanup", py::capsule(cleanup_callback));
```

Am I right, that this solves my issue? Is `cleanup_callback` called before the interpreter goes down but after all my python objects are destructed?

**Comment by wjakob** [comment]
Yes, this should solve your issue. Can you give it a try and report back?

**Comment by thorink** [comment]
I'm on it.

**Comment by thorink** [comment]
It seems to work, though I needed to modify the example a bit because clang gave me a `error: non-const lvalue reference to type 'pybind11::object' cannot bind to a temporary of type 'py::capsule'`:

```c++
auto cleanup_callback = []() {
    // perform cleanup here -- this function is called with the GIL held
};

// does not compile:
// m.add_object("_cleanup", py::capsule(cleanup_callback));

// works:
py::capsule cap(cleanup_callback);
m.add_object("_cleanup", cap);
```

I'll set up an documentation fix MR.

**Comment by wjakob** [comment]
@thorink: can you check if the code works with the latest master version that includes patch aa1b316?

**Comment by thorink** [comment]
I've tested it with a patched v2.1 and it worked for me.
